hashtable.t

```diff
@@ -0,0 +1,509 @@
+	end
+
+	-- Inserts a bucket into the provided hashtable.
+	-- Returns a result containing an InsertData or an error code 
+	local terra insert_bucket(bucket: BucketType, metadata_array: &uint8, buckets_array: &BucketType, capacity: uint): InsertResult 
+		var hash_info, index = hash_probe(bucket.key, metadata_array, buckets_array, capacity):unwrap()
+		var old_metadata = metadata_array[index]
+
+		metadata_array[index] = hash_info.h2
+		buckets_array[index] = bucket
+
+		return InsertResult.ok(InsertData {old_metadata, index, hash_info})
+	end
+
+	local terra find_next_power_of_two(m: uint): uint
+		-- TODO: This only works for 32 bit integers. Need to modifiy for 64 bit if the int size is 64 bit.
+		m = m - 1
+		m = m or (m >> 1)
+		m = m or (m >> 2)
+		m = m or (m >> 4)
+		m = m or (m >> 8)
+		m = m or (m >> 16)
+		m = m + 1
+		return m
+	end
+
+	-- Checks to see if the provided capacity and size indicate that the table needs to resize
+	-- Implemented as a function so that it can optimized in the future
+	local terra is_at_capacity(capacity: uint, size: uint): bool
+		return capacity == size
+	end
+
+	terra HashTable:init()
+		var initial_capacity = GroupLength
+		var calloc_result = table_calloc(initial_capacity)
+		
+		if calloc_result:is_ok() then
+			var opaque_ptr, metadata, buckets = calloc_result.ok
+			
+			self:_init {
+				capacity = initial_capacity,
+				size = 0,
+				opaque_ptr = opaque_ptr,
+				metadata = metadata,
+				buckets = buckets
+			}
+		end
+	end
+
+	terra HashTable:destruct()
+		if self.opaque_ptr ~= nil then
+			[Alloc]:free_raw(self.opaque_ptr)
+			CStr.memset(self, 0, sizeof(HashTable))
+		end
+	end
+
+	-- Resizes the hashtable to be at least the size of the requested_capacity.
+	-- Returns 0 if there was no error. 
+	terra HashTable:reserve(requested_capacity: uint): uint
+		-- If the requested_capacity is lower than self.capacity, return
+		if requested_capacity < self.capacity then
+			return M.Errors.ValueError
+		end
+
+		var new_capacity = find_next_power_of_two(requested_capacity) 
+		var new_opaque, new_metadata, new_buckets = table_calloc(new_capacity):unwrap()
+
+		-- Iterate through the old data and rehash existing entries
+		for i = 0, self.capacity do
+			if metadata_holds_hash(self.metadata[i]) then
+				var bucket = self.buckets[i]
+				var insert_result = insert_bucket(bucket, new_metadata, new_buckets, new_capacity)
+
+				if insert_result:is_err() then
+					[Alloc]:free_raw(new_opaque)
+					new_opaque = nil
+					new_metadata = nil
+					new_buckets = nil
+					return insert_result.err
+				end
+			end
+		end
+
+		[Alloc]:free_raw(self.opaque_ptr)
+		self.opaque_ptr = new_opaque
+		self.metadata = new_metadata
+		self.buckets = new_buckets
+		self.capacity = new_capacity
+
+		return 0
+	end
+
+	terra HashTable:has(key: KeyType): bool
+		var result = hash_probe(key, self.metadata, self.buckets, self.capacity) 
+
+		if result:is_ok() and metadata_holds_hash(self.metadata[result.ok._1]) then
```

Comment:
this line isn't very self-explanatory. Could the field maybe be named? or commented here?